service/src/main/java/dk/dbc/scan/service/ScanLogic.java

```diff
@@ -0,0 +1,201 @@
+    }
+
+    /**
+     * Callback when an asynchronous call is done
+     */
+    private synchronized void inFlightDone() {
+        hitCountInFlight.decrementAndGet();
+        notify();
+    }
+
+    /**
+     * Wait for all asynchronous calls to be completed
+     *
+     * @throws InterruptedException If the thread is interrupted
+     */
+    private synchronized void awaitAllInFlight() throws InterruptedException {
+        while (hitCountInFlight.get() != 0) {
+            wait();
+        }
+    }
+
+    /**
+     * Get the number of terms that are valid for the result
+     *
+     * @return number of terms that are ready to be returned
+     */
+    private int goodTermCount() {
+        return (int) termsFound.stream()
+                .filter(ScanResponse.Term::hasTerms)
+                .count();
+    }
+
+    /**
+     * Build result for scan
+     *
+     * @param register   The register to scan in
+     * @param term       The term to scan with
+     * @param cont       If current term shouldn't be included. This has the
+     *                   side-effect that if term is included, then it is normalized, by solr
+     *                   using a field-analysis request
+     * @param count      Number of terms wanted in result
+     * @param agencyId   Agency that performs the request
+     * @param profile    The profile that is used by the agency
+     * @param trackingId Tracking
+     * @return Response to end-user
+     * @throws IOException          In case of communication errors with solr
+     * @throws SolrServerException  In case of request syntax errors or invalid
+     *                              register name
+     * @throws InterruptedException If something is timing out, or the process
+     *                              is being shut down
+     */
+    public ScanResponse.Result scan(String register, String term, boolean cont, int count, String agencyId, String profile, String trackingId) throws IOException, SolrServerException, InterruptedException {
+        String scanRegister = register + "_" + agencyId + "_" + profile;
+        hitCountInFlight = new AtomicInteger(0);
+        termsFound = new ArrayList<>();
+
+        if (!cont) {
+            term = solr.normalize(register, term);
+            log.debug("normalized = {}", term);
+        }
+
+        String filterQuery = psCache.filterQueryFor(agencyId, profile, trackingId);
+        log.debug("filterQuery = {}", filterQuery);
+
+        Iterator<String> terms = EMPTY_ITERATOR;
+        while (goodTermCount() < count) {
+            if (!terms.hasNext()) {
+                if (term == null) {// We've reached end of index
+                    log.debug("We've reached the end of the index");
+                    break;
+                }
+                int toGo = count - goodTermCount();
+                int fetch = toGo + 5 + toGo / 8; // extra terms to fetch, so that zero hits can be skipped
```

Comment:
Selvom der er en kommentar virker det her stadig som tal magi for mig... Forstår ikke hvad der foregår

Reply:
Der skal heltes et antal termer, som hitcount valideres. Hvis dette hitcount er 0 smides den væk. Derfor skal der hentes flere termer end der skal returneres, da enkelte risikerer at blive smidt væk.
Når et dokument i SolR bliver slettet forsvinder nøglerne ikke fra term-indekset før det bliver merget. Derfor kan der være ord i term-indekset som giver 0 hit (f.eks. hvis en tastefejl i titlen rettes).
En god kommentar modtages gerne.
+5 +12.5% er bare en mavefornemmelse (sikkert alt for høj)